src/state/workload.rs

```diff
@@ -632,14 +633,83 @@ pub struct WorkloadStore {
     // is simpler (and only requires a channelsize of 1)
     insert_notifier: Sender<()>,
 
-    /// byAddress maps workload network addresses to workloads
-    by_addr: HashMap<NetworkAddress, Arc<Workload>>,
-    /// byUid maps workload UIDs to workloads
+    /// by_addr maps workload network addresses to workloads
+    by_addr: HashMap<NetworkAddress, WorkloadByAddr>,
+    /// by_uid maps workload UIDs to workloads
     pub(super) by_uid: HashMap<Strng, Arc<Workload>>,
     // Identity->Set of UIDs. Only stores local nodes
     node_local_by_identity: HashMap<WorkloadIdentity, HashSet<Strng>>,
 }
 
+#[derive(Debug)]
+/// WorkloadByAddr is a small wrapper around a single or multiple Workloads
+/// We split these as in the vast majority of cases there is only a single one, so we save vec allocation.
+enum WorkloadByAddr {
+    Single(Arc<Workload>),
+    Many(Vec<Arc<Workload>>),
+}
+
+impl WorkloadByAddr {
+    // insert adds the workload
+    pub fn insert(&mut self, w: Arc<Workload>) {
+        match self {
+            WorkloadByAddr::Single(workload) => {
+                *self = WorkloadByAddr::Many(vec![workload.clone(), w]);
+            }
+            WorkloadByAddr::Many(v) => {
+                v.push(w);
+            }
+        }
+    }
+    // remove_uid mutates the address to remove the workload referenced by the UID.
+    // If 'true' is returned, there is no workload remaining at all
+    pub fn remove_uid(&mut self, uid: Strng) -> bool {
+        match self {
+            WorkloadByAddr::Single(wl) => {
+                // Remove it if the UID matches, else do nothing
+                wl.uid == uid
+            }
+            WorkloadByAddr::Many(ws) => {
+                ws.retain(|w| w.uid != uid);
+                match ws.as_slice() {
+                    [] => true,
+                    [wl] => {
+                        // We now have one workload, transition to Single
+                        *self = WorkloadByAddr::Single(wl.clone());
+                        false
+                    }
+                    // We still have many. We removed already so no need to do anything
+                    _ => false,
+                }
+            }
+        }
+    }
+    pub fn get(&self) -> Arc<Workload> {
```

Comment:
Does it make sense to add another criteria here for the network of the workload? Thinking of a scenario where the control plane is sending WDS to two clusters that have overlapping IP addresses. This isn't supported now, so the correct answer may be "modify this if/when we add support for it", but just throwing it out there

Reply:
The key into the map is already NetworkAddress so its already handled

Reply:
Ah right `NetworkAddress` has the `network/` prefix. Makes sense